tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmArtificialInstInterpreter.kt

```diff
@@ -280,52 +296,194 @@ class TvmArtificialInstInterpreter(
+                        newStmt(
+                            TsaArtificialOnOutMessageHandlerCallInst(
+                                computePhaseResult = stmt.computePhaseResult,
+                                actionPhaseResult = result,
+                                location = lastStmt.location,
+                                sentMessages = messages,
+                                messageOrderNumber = 0,
+                            ),
+                        )
+                    }
+                }
 
-        val commitedState =
-            scope.calcOnState {
-                lastCommitedStateOfContracts[currentContract]
+                is ActionHandlingResult.RealFailure -> {
+                    this.calcOnState {
+                        val failure =
+                            TvmFailure(
+                                actionsHandlingResult.failure,
+                                TvmFailureType.UnknownError,
+                                phase,
+                                pathNode,
+                            )
+                        calcOnState {
+                            isExceptional = isExceptional || oldIsExceptional
+                        }
+                        newStmt(TsaArtificialExitInst(stmt.computePhaseResult, failure, lastStmt.location))
+                    }
+                }
+
+                is ActionHandlingResult.SoftFailure -> {
+                    this.calcOnState {
+                        val failure =
+                            TvmResult.TvmSoftFailure(
+                                actionsHandlingResult.failure,
+                                phase,
+                            )
+                        calcOnState {
+                            isExceptional = isExceptional || oldIsExceptional
+                        }
+                        newStmt(TsaArtificialExitInst(stmt.computePhaseResult, failure, lastStmt.location))
+                    }
+                }
             }
+        }
+    }
 
+    private fun visitParseActionInst(
+        scope: TvmStepScopeManager,
+        stmt: TsaArtificialActionParseInst,
+    ) {
         // temporarily remove isExceptional mark for further processing
         val oldIsExceptional = scope.calcOnState { isExceptional }
         scope.doWithState {
             isExceptional = false
         }
+        val (head, tail) =
+            stmt.yetUnparsedActions.splitHeadTail() ?: return run {
+                transactionInterpreter.resolveMessageReceivers(
+                    scope,
+                    stmt.parsedAndPreprocessedActions,
+                ) { parsingResult ->
+                    doWithState {
+                        newStmt(
+                            TsaArtificialHandleMessagesCostInst(stmt.computePhaseResult, stmt.location, parsingResult),
+                        )
+                    }
+                }
+            }
+        val tmpStmt = stmt.copy(yetUnparsedActions = tail)
```

Comment:
Isn't this `nextStmt`?

Reply:
No, it isn't, but the name choice is definitely poor. I'll try to fix it